server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -4436,6 +4475,61 @@ private SnapshotsInProgress createSnapshot(
         }
     }
 
+    private Collection<LongWithAttributes> getSnapshotsInProgress() {
+        final ClusterState currentState = clusterService.state();
+        // Only the master should report on snapshots-in-progress
+        if (currentState.nodes().isLocalNodeElectedMaster() == false) {
+            return List.of();
+        }
+        final SnapshotsInProgress snapshotsInProgress = SnapshotsInProgress.get(currentState);
+        final List<LongWithAttributes> snapshotsInProgressMetrics = new ArrayList<>();
+        currentState.metadata().projects().forEach((projectId, project) -> {
+            final RepositoriesMetadata repositoriesMetadata = RepositoriesMetadata.get(project);
+            if (repositoriesMetadata != null) {
+                repositoriesMetadata.repositories().forEach(repository -> {
+                    int snapshotCount = snapshotsInProgress.forRepo(projectId, repository.name()).size();
+                    snapshotsInProgressMetrics.add(
+                        new LongWithAttributes(snapshotCount, SnapshotMetrics.createAttributesMap(projectId, repository))
+                    );
+                });
+            }
+        });
+        return snapshotsInProgressMetrics;
+    }
+
+    private Collection<LongWithAttributes> getShardsByState() {
+        final ClusterState currentState = clusterService.state();
+        // Only the master should report on shards-by-state
+        if (currentState.nodes().isLocalNodeElectedMaster() == false) {
+            return List.of();
+        }
+        final SnapshotsInProgress snapshotsInProgress = SnapshotsInProgress.get(currentState);
+        final List<LongWithAttributes> shardsByState = new ArrayList<>();
+        final ObjectIntMap<ShardState> shardCounts = new ObjectIntHashMap<>(TRACKED_SHARD_STATES.size());
+        currentState.metadata().projects().forEach((projectId, project) -> {
+            final RepositoriesMetadata repositoriesMetadata = RepositoriesMetadata.get(project);
+            if (repositoriesMetadata != null) {
+                for (RepositoryMetadata repository : repositoriesMetadata.repositories()) {
+                    TRACKED_SHARD_STATES.forEach(shardState -> shardCounts.put(shardState, 0));
+                    for (SnapshotsInProgress.Entry snapshot : snapshotsInProgress.forRepo(projectId, repository.name())) {
+                        for (ShardSnapshotStatus shardSnapshotStatus : snapshot.shards().values()) {
+                            if (shardCounts.containsKey(shardSnapshotStatus.state())) {
+                                shardCounts.put(shardSnapshotStatus.state(), shardCounts.get(shardSnapshotStatus.state()) + 1);
```

Comment:
suggest using `com.carrotsearch.hppc.ObjectIntMap#addTo` (saves looking up the entry twice)

Reply:
Fixed in ba62ff63e4dd914d575c8ffd48c2a8c4db1312cb